pkg/depth/buffer.go

```diff
@@ -155,48 +160,72 @@ func (b *Buffer) AddUpdate(o types.SliceOrderBook, firstUpdateID int64, finalArg
 			b.logger.Info("try fetching the snapshot due to no snapshot")
 			go b.tryFetch()
 		})
-		b.mu.Unlock()
+
 		return nil
 	}
 
 	// if it's ready, then we have the snapshot, we can push the update
 
 	// skip older events
 	if u.FinalUpdateID <= b.finalUpdateID {
-		b.logger.Infof("the final update id %d of event is less than equal to the final update id %d of the snapshot, skip", u.FinalUpdateID, b.finalUpdateID)
-		b.mu.Unlock()
+		b.logger.Infof("the final update id %d of event is less than equal to the final update id %d of the snapshot, skip",
+			u.FinalUpdateID, b.finalUpdateID)
 		return nil
 	}
 
 	// if there is a missing update, we should reset the snapshot and re-fetch the snapshot
-	if u.FirstUpdateID > b.finalUpdateID+1 {
-		// drop the prior updates in the buffer since it's corrupted
-		b.buffer = []Update{u}
-		b.resetSnapshot()
-		b.once.Reset()
-		b.once.Do(func() {
-			b.logger.Info("try fetching the snapshot due to missing update")
-			go b.tryFetch()
-		})
-
-		b.mu.Unlock()
-		b.EmitReset()
-
-		return fmt.Errorf("found missing update between finalUpdateID %d and firstUpdateID %d, diff: %d",
-			b.finalUpdateID+1,
-			u.FirstUpdateID,
-			u.FirstUpdateID-b.finalUpdateID)
+	if err := b.checkMissingUpdates(u); err != nil {
+		return err
 	}
 
 	b.logger.Debugf("depth update id %d -> %d", b.finalUpdateID, u.FinalUpdateID)
 	b.finalUpdateID = u.FinalUpdateID
 	b.EmitPush(u)
 
-	b.mu.Unlock()
+	return nil
+}
+
+func (b *Buffer) handleFetchSignal(u Update) {
+	b.logger.Info("fetch signal received")
+	b.buffer = append(b.buffer, u)
+	b.resetSnapshot()
+	b.once.Reset()
+	b.once.Do(func() {
+		b.logger.Info("try fetching the snapshot due to fetch signal received")
+		go b.tryFetch()
+	})
+}
 
+func (b *Buffer) checkMissingUpdates(u Update) error {
+	if b.isFutures {
+		previousUpdateID := b.finalUpdateID
+		if previousUpdateID != 0 && u.PreviousUpdateID != previousUpdateID {
+			b.handleMissingUpdate(u)
+			return fmt.Errorf("found missing update, new event's previousUpdateID: %d not equal previous event's finalUpdateID: %d",
+				u.PreviousUpdateID, previousUpdateID)
+		}
+	} else {
+		if u.FirstUpdateID > b.finalUpdateID+1 {
+			b.handleMissingUpdate(u)
+			return fmt.Errorf("found missing update between finalUpdateID %d and firstUpdateID %d, diff: %d",
+				b.finalUpdateID+1, u.FirstUpdateID, u.FirstUpdateID-b.finalUpdateID)
+		}
+	}
 	return nil
 }
 
+func (b *Buffer) handleMissingUpdate(u Update) {
+	// drop the prior updates in the buffer since it's corrupted
+	b.buffer = []Update{u}
+	b.resetSnapshot()
+	b.once.Reset()
+	b.once.Do(func() {
+		b.logger.Info("try fetching the snapshot due to missing update")
+		go b.tryFetch()
+	})
+	b.EmitReset()
```

Comment:
EmitReset needs to be called after the mutex Unlock in order to avoid nested locking.

In your case, when calling this method, the mutex is still in the locked mode, did you verify this behavior? does it cause other effect?